java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -338,14 +340,32 @@ class DatabaseImpl(
                 }
             }
 
-            if (isCollection) {
+            when (isCollection) {
+                FieldClass.Collection -> {
                 // Ensure we create the collection even if the element to add is null.
-                val collection = collections.getOrPut(fieldName) { mutableSetOf() }
-                value?.let { x -> collection.add(x) }
-            } else {
-                singletons[fieldName] = value
+                    val collection = collections.getOrPut(fieldName) { mutableSetOf() }
+                    value?.let { x -> collection.add(x) }
+                }
+                FieldClass.List -> {
+                    val list = lists.getOrPut(fieldName) { mutableListOf() }
+                    if (typeId > LARGEST_PRIMITIVE_TYPE_ID) {
+                        listTypes.put(fieldName, FieldType.EntityRef(getSchemaHash(typeId, db)))
```

Comment:
with getSchemaHash you are potentially doing a database lookup for each value in each list.
Maybe we could skip this step if listtypes already contains this fieldName?
or add a TODO to address later

Reply:
that's a good idea, I'll implement it now.